src/grain/grain.js

```diff
@@ -108,3 +108,36 @@ export function format(
   }
   return digits.join("") + suffix;
 }
+
+/**
+ * Multiply a grain amount by a floating point number.
+ *
+ * Use this method when you need to multiply a grain balance by a floating
+ * point number, e.g. a ratio.
+ *
+ * Note that this method is imprecise. It is not safe to assume, for example,
+ * that `multiply(g, 1/3) + multiply(g, 2/3) === g` due to loss of precision.
+ * However, the errors will be small in absolute terms (i.e. tiny compared to
+ * one full grain).
+ *
+ * See some messy analysis of the numerical errors here:
+ * https://observablehq.com/@decentralion/grain-arithmetic
+ */
+export function multiplyFloat(grain: Grain, num: number): Grain {
+  return BigInt(Math.floor(Number(grain) * num));
```

Comment:
Hmm… this conversion is a bit awkward considering that `grain` will
exceed `Number.MAX_SAFE_INTEGER` (2^53 − 1 ≈ 10^16 < 10^18). This means
that it’s not even the case that `multiplyFloat(g, 1) === g`, even for
such pedestrian values of `g` as `ONE + 1n`:

```javascript
> g = (10n ** 18n) + 1n;
1000000000000000001n
> BigInt(Math.floor(Number(g) * 1))
1000000000000000000n
```

(And, in fact, all values between `ONE - 64n` and `ONE + 64n` inclusive
have the same image under `(g) => multiplyFloat(g, 1)`.)

That is, this function is approximate not just in the floating point
argument but in the grain argument itself.

This deficit is not fundamental. We can do the calculation entirely in
bigint space, destructuring the float into its integer components; quick
sketch:

```javascript
function mul(grain: Grain, fac: number) {
  const {sgn, exponent, mantissa} = decompose(fac);
  // all small integers:
  //   - sgn: either -1 or +1
  //   - exponent: between -1023 and 1024 (assuming normal)
  //   - mantissa: between 0 and 2^51 - 1 (safe by construction)
  // such that fac = sgn * 2^exponent * (1 + mantissa / 2^52)
  // so
  //     fac * 2^1023 * 2^52 = sgn * 2^(exponent+1023) * (2^52 + mantissa)
  // is a product of integers---
  const bigFac =
    BigInt(sgn) *
    2n ** (BigInt(exponent) + 1023n) *
    (2n ** 52n + BigInt(mantissa));
  // ---and thus the bigint arithmetic is correct to within 1ulp (could
  // be 0.5ulp, but preserving floor semantics per your code):
  return (grain * bigFac) / 2n ** (1023n + 52n);
}
```

You need a bit of care for subnormals, but honestly it would just be
fine to throw on them and not support multiplying by `fac`s of magnitude
smaller than `1 / Number.MAX_VALUE` ≈ 5.5e-309. (Lots of floating point
code doesn’t support subnormals.)

Note that in this implementation we only ever convert from floating
point to bigint—a lossless conversion, even if the floating point input
is itself approximate. We never convert from bigint to floating point,
which is inherently lossy. Thus, we have sane properties at least in the
first argument, like:

  - `mul(g, 1) === g`
  - `mul(g1, f) + mul(g2, f) === mul(g1 + g2, f)`
  - `mul(k * g, f) === k * mul(g, f)` (for `k` a `BigInt`)

…and moreover it is immediate that you have those properties because
`grain` is only used in `mul` as a factor in a bigint product.

Reply:
A high-level summary of the second half of this comment, starting from
“This deficit is not fundamental”, is: “Every (finite) floating point
number exactly represents a ratio of two integers. If `fac = p / q`,
then just take `g * BigInt(p) / BigInt(q)` to compute the exact result.”

Reply:
Maybe an even higher-level rule of thumb: 64-bit floats only have about
16 significant figures (decimal) of precision. You’re trying to create a
fixed-point type whose scaling factor is 10^18. Thus, you can *never*
take `Number(grain)` without losing data.

Reply:
Got it. As discussed offline, let's use your fixes. I'm going to merge this one to unblock development, please send a followon PR which fixes the implementation.